Core/Spells: Fixed SPELL_AURA_MOD_CASTING_SPEED_NOT_STACK stacking

Keader opened the pull request:
**Changes proposed:**

-  Let only more powerful effect 

Change based in spell name + this 2.4.0 patch note:
```
Power Infusion: Infuses the target with power, increasing their spell haste by 20% and reducing the mana cost of all spells by 20%. Lasts for 15 seconds. This will not stack with other haste effects, such as Heroism, Bloodlust, or Icy Veins.
```
https://wowwiki.fandom.com/wiki/Patch_2.4.0

**Target branch(es):** 3.3.5

**Tests performed:** Builded and tested in game

Keader committed:
Core/Spells: Fixed SPELL_AURA_MOD_CASTING_SPEED_NOT_STACK stacking

P-Kito commented on src/server/game/Spells/Auras/SpellAuraEffects.cpp:
```diff
@@ -3901,6 +3901,24 @@ void AuraEffect::HandleModCastingSpeed(AuraApplication const* aurApp, uint8 mode
+        Unit::AuraEffectList const& castingSpeedNotStack = target->GetAuraEffectsByType(SPELL_AURA_MOD_CASTING_SPEED_NOT_STACK);
+        for (AuraEffect const* aurEff : castingSpeedNotStack)
+        {
+            if (aurEff != this && abs(strongEffect) < abs(aurEff->GetAmount()))
```
Hey Keader <3 Question: What would happen if there is -70 effect and a +50 effect. Would that ultimately mean that the -70 will be a +70 effect?

Keader replied:
Nope, we still use non-abs value.
It's a copy of SpellGroup system.

ForesterDev commented:
Why not use SPELL_GROUP_STACK_RULE_EXCLUSIVE_HIGHEST implementation for this? Mayve improve HandleModCastingSpeed function to use SpellGroups?

Keader commented:
Because we always have ```SPELL_GROUP_STACK_RULE_EXCLUSIVE_HIGHEST``` in this case.
This PR is  basically a copy  of ```SPELL_GROUP_STACK_RULE_EXCLUSIVE_HIGHEST``` implementation, just dont get any DB value.
To work with SpellGroups, we need add 200+ spells in DB, i dont wanna do it. Since i belive that all should not stack (so all should have same rule).

Keader commented:
Hmm, i'm missing misc value...
Need search a little more.

Keader committed:
Some spells need stack, so we need check SchoolMask too

Keader commented:
Some spells like Nature's Grace, Air totem, Celestial focus need stack.
Looking into Spellwork, looks like SchoolMask need be checked for those cases.

Jildor commented:
Info about this: http://talesofapriest.com/re-revisiting-power-infusion-in-wrath/

btw: https://www.mmo-champion.com/threads/692845-Arcane-Power-Power-Infusion-Fail
Power Infusion and Arcane Power shouldn't stack? (Bugged too I think 😄 )

Keader commented:
Since Arcane Power is not SPELL_AURA_MOD_CASTING_SPEED_NOT_STACK will not fixed by this PR.
But i think that SpellGroups can handle with it. I will check later.

Jildor commented:
Yep I kno, I put here because I found searching the PR info 😄 
Yeah spellgroup should fix this ;)

Keader committed:
Added support to SpellGroups

Keader committed:
sql

Keader commented:
Changed scope.
Eradication should stack with BL/Heroism, so break my theory about SPELL_AURA_MOD_CASTING_SPEED_NOT_STACK.
So fixed original case (BL/Heroism vs Power Infusion) and added support to SpellGroup in HandleModCastingSpeed.

Keader committed:
Ops, forgot icy veins

jackpoz commented on src/server/game/Spells/Auras/SpellAuraEffects.cpp:
```diff
@@ -3901,6 +3901,13 @@ void AuraEffect::HandleModCastingSpeed(AuraApplication const* aurApp, uint8 mode
     }
 
+    int32 spellGroupVal = target->GetHighestExclusiveSameEffectSpellGroupValue(this, GetAuraType());
+    if (abs(spellGroupVal) >= abs(GetAmount()))
```
I would prefer to use std::abs() rather than abs()

Keader replied:
all other handlers is using abs()
Need geral change then 💃

lost-illusion commented:
@Keader Hi. I have question about IV, after this change you can't use IV while Bloodlust is active.
But...
https://youtu.be/rdg6U9SD0B8?t=203
https://www.dslreports.com/forum/r23100096-Is-it-better-to-extend-or-stack-haste-buffs
https://www.wowhead.com/forums&topic=133764
https://www.ownedcore.com/forums/world-of-warcraft/world-of-warcraft-guides/281258-how-successful-mage.html
And etc, say "You can use IV + BL, but you loose part of haste because 1 sec gcd". I think this PR may be wrong.

Jildor commented:
https://github.com/TrinityCore/TrinityCore/commit/af874a211c0f064bd685cddeded3ca7d6d690e8f

lost-illusion commented:
Oh, my bad. Sorry.
